### PR TITLE
Fractional multip

### PR DESCRIPTION
**Pull request by sambitmishra98**
This allows the number of smoothing operations in the configuration file to be a float, instead of an integer. Number of smoothing operations are chosen using a uniform distribution, such that the average of smoothing operations performed matches with the input float number in the configuration.

### CODE REVIEW

File: pyfr/integrators/dual/pseudo/multip.py [hunk 1]
```diff
@@ -23,6 +24,8 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg,
 
         # Get the multigrid cycle
         self.cycle, self.csteps = zip(*cfg.getliteral(mgsect, 'cycle'))
+        random.seed(cfg.getint(mgsect, 'seed', 0))
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:27]: Changing the global seed is not appropriate.  Instead create a NumPy generator object.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 2]
```diff
@@ -288,6 +297,7 @@ def pseudo_advance(self, tcurr):
         self.tcurr = tcurr
 
         for i in range(self._maxniters):
+            csteps = tuple(self.wround(cstepf) for cstepf in cstepsf)
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:300]: Use choices [np.floor(c), np.ceil(c)] being the two choices with probabilities of [c % 1 and 1 - c % 1]

File: pyfr/integrators/dual/pseudo/multip.py [hunk 3]
```diff
@@ -23,6 +25,8 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg,
 
         # Get the multigrid cycle
         self.cycle, self.csteps = zip(*cfg.getliteral(mgsect, 'cycle'))
+        self.rng = np.random.default_rng(cfg.getint(mgsect, 'seed', 0))
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:28]: I would not expose a seed.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 4]
```diff
@@ -275,10 +279,10 @@ def _mg_regidx(self):
                                  ' self.level == self._order')
 
         return self.pintg._aux_regidx
-
+
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:282]: Whitespace.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 5]
```diff
@@ -288,6 +292,12 @@ def pseudo_advance(self, tcurr):
         self.tcurr = tcurr
 
         for i in range(self._maxniters):
+
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:295]: Whitespace.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 6]
```diff
@@ -288,6 +292,12 @@ def pseudo_advance(self, tcurr):
         self.tcurr = tcurr
 
         for i in range(self._maxniters):
+            
+            # Choose either ⌊c⌋ and ⌈c⌉ in a way that the average is c
+            csteps = tuple(self.rng.choice([np.floor(c), np.ceil(c)],
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:297]: Use a list rather than a tuple.

File: pyfr/integrators/dual/pseudo/multip.py (hunk 3 above)
Comment (FreddieWitherden) [reply pyfr/integrators/dual/pseudo/multip.py:28]: Also I'd make this a private member with a suitable name.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 7]
```diff
@@ -288,6 +292,11 @@ def pseudo_advance(self, tcurr):
         self.tcurr = tcurr
 
         for i in range(self._maxniters):
+
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:295]: Remove newline.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 8]
```diff
@@ -288,6 +292,11 @@ def pseudo_advance(self, tcurr):
...
+            # Choose either ⌊c⌋ and ⌈c⌉ in a way that the average is c
+            csteps = [self._fgen.choice([np.floor(c), np.ceil(c)], 
+                                       p=[c % 1, 1 - c % 1]) for c in cstepsf]
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:298]: Bad indentation. (Saying `fgen = self._fgen` up top may help clean things up.)

File: pyfr/integrators/dual/pseudo/multip.py (hunk 7 above)
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:295]: Remove newline.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 9]
```diff
@@ -288,6 +292,11 @@ def pseudo_advance(self, tcurr):
         self.tcurr = tcurr
 
         for i in range(self._maxniters):
+
+            # Choose either ⌊c⌋ and ⌈c⌉ in a way that the average is c
+            csteps = [int(self._fgen.choice([np.floor(c), np.ceil(c)],
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:297]: Bad indentation.

**Comment by FreddieWitherden** [comment]
Also, for example purposes please update the incompressible cylinder example to use a fractional cycle and add appropriate documentation.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 10]
```diff
@@ -23,6 +25,8 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg,
 
         # Get the multigrid cycle
         self.cycle, self.csteps = zip(*cfg.getliteral(mgsect, 'cycle'))
+        self._fgen = np.random.default_rng(0)
```
Comment (WillTrojak) [inline pyfr/integrators/dual/pseudo/multip.py:28]: For reproducibility it might be worth being explicit about the generator

Comment (WillTrojak) [reply pyfr/integrators/dual/pseudo/multip.py:28]: For example
```python
seed = self.cfg.getint('solver', 'seed', 1234)
self._fgen = np.random.Generator(np.random.MT19937(seed))
```

File: pyfr/integrators/dual/pseudo/multip.py [hunk 11]
```diff
@@ -278,16 +282,21 @@ def _mg_regidx(self):
...
             self.pintgs[l]._dt = self._dt
             self.pintgs[l].stepper_coeffs = self.stepper_coeffs
 
         self.tcurr = tcurr
+        fgen = self._fgen
 
         for i in range(self._maxniters):
+            # Choose either ⌊c⌋ and ⌈c⌉ in a way that the average is c
+            csteps = [int(self._fgen.choice([np.floor(c), np.ceil(c)], 
+                          p=[1 - c % 1, c % 1])) for c in cstepsf]
```
Comment (WillTrojak) [inline pyfr/integrators/dual/pseudo/multip.py:298]: Would this be cleaner?
```
    csteps = [int(np.floor(c) if fgen.random() < 1 - c % 1 else np.ceil(c)) for c in cstepsf]
```

**Review by WillTrojak — COMMENTED** [review]
Just a couple points

File: pyfr/integrators/dual/pseudo/multip.py (hunk 10 above)
Comment (sambitmishra98) [reply pyfr/integrators/dual/pseudo/multip.py:28]: Freddie said not to expose the seed: https://github.com/PyFR/PyFR/pull/362#discussion_r1296828282

File: pyfr/integrators/dual/pseudo/multip.py (hunk 11 above)
Comment (sambitmishra98) [reply pyfr/integrators/dual/pseudo/multip.py:298]: Ok. Rearranging a bit:
```
csteps = [int(np.floor(c) if fgen.random() > c % 1 else np.ceil(c)) for c in cstepsf]
```

Or the below line, which runs 2 times faster and fits in the 80 lines limit, but lesser readable.
```
csteps = [int(c + (fgen.random() < c % 1)) for c in cstepsf]
```

Comment (sambitmishra98) [reply pyfr/integrators/dual/pseudo/multip.py:298]: I think we can use the second one. I have anyways commented above that line what the line does.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 12]
```diff
@@ -23,6 +25,8 @@ def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg,
 
         # Get the multigrid cycle
         self.cycle, self.csteps = zip(*cfg.getliteral(mgsect, 'cycle'))
+        self._fgen = np.random.Generator(np.random.MT19937(0))
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:28]: Prefer https://numpy.org/doc/stable/reference/random/bit_generators/pcg64.html#numpy.random.PCG64

File: pyfr/integrators/dual/pseudo/multip.py [hunk 13]
```diff
@@ -278,16 +282,20 @@ def _mg_regidx(self):
...
         # Set time step and current stepper coefficients for all levels
         for l in self.levels:
             self.pintgs[l]._dt = self._dt
             self.pintgs[l].stepper_coeffs = self.stepper_coeffs
 
         self.tcurr = tcurr
+        fgen = self._fgen
 
         for i in range(self._maxniters):
+            # Choose either ⌊c⌋ and ⌈c⌉ in a way that the average is c
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:296]: or

File: pyfr/integrators/dual/pseudo/multip.py [hunk 14]
```diff
@@ -278,16 +282,20 @@ def _mg_regidx(self):
...
-        cycle, csteps = self.cycle, self.csteps
+        cycle, cstepsf = self.cycle, self.csteps
 
         # Set time step and current stepper coefficients for all levels
         for l in self.levels:
             self.pintgs[l]._dt = self._dt
             self.pintgs[l].stepper_coeffs = self.stepper_coeffs
 
         self.tcurr = tcurr
+        fgen = self._fgen
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py:293]: Move this to the start of the function

**Comment by FreddieWitherden** [comment]
Are you okay to also submit a PR to the test cases and also update the documentation?

**Pull request merged**